### PR TITLE
Fix heuristic scheduling for UAG

### LINKED ISSUE #34803: Heuristic scheduling for UAG
https://github.com/huggingface/transformers/blob/5de58d595547f57dd4df2f3572158eb9cbdd24c9/src/transformers/generation/candidate_generator.py#L238-L261

When using [Universal Assisted Generation (UAG)](https://huggingface.co/docs/transformers/en/generation_strategies#universal-assisted-decoding), draft and target vocabularies are different.

In `heuristic` scheduling, the `update_candidate_strategy` function adjusts the maximum number of draft tokens for the next iteration by verifying whether num_matches equals self.num_assistant_tokens.

However, in UAG:
- num_matches: Represents the number of matched tokens according to the **target** vocabulary.
- self.num_assistant_tokens: Reflects the number of tokens generated by the **draft** model, based on the **draft** vocabulary.

### PR DESCRIPTION
**Pull request by jmamou**
# What does this PR do?
Fix #34803

The number of matching tokens according to the draft vocabulary is missing in [candidate_generator.py](https://github.com/huggingface/transformers/blob/main/src/transformers/generation/candidate_generator.py). To fix the [issue](https://github.com/huggingface/transformers/issues/34803), we propose to modify [update_candidate_strategy](https://github.com/huggingface/transformers/blob/main/src/transformers/generation/candidate_generator.py#L254) by comparing `num_matches` (the number of matched tokens according to the **target** vocabulary) with `len(scores[0]) - 1` that represents the number of tokens generated by the **draft** model according to the **target** vocabulary.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
 @gante @ArthurZucker @zucchini-nlp

### CODE REVIEW

**Review by zucchini-nlp — COMMENTED** [review]
@jmamou thanks for fixing! Quick clarification, why is there a `-1`, shouldn't `scores` reflect the length of generated tokens?

**Comment by jmamou** [comment]
> @jmamou thanks for fixing! Quick clarification, why is there a `-1`, shouldn't `scores` reflect the length of generated tokens?

@zucchini-nlp 

`scores` contains also the additional token produced by the target model independently of the draft generation process as  explained [here](https://github.com/huggingface/transformers/blob/9470d6532436e9db2951a196effd6f8841befb76/src/transformers/generation/utils.py#L4253-L4255)

File: src/transformers/generation/candidate_generator.py
Comment (zucchini-nlp) [inline src/transformers/generation/candidate_generator.py:260]: nit: we can put comment on line above so it doesn't get formatted like this

**Review by zucchini-nlp — APPROVED** [review]
Ah, I see, thanks for explaining. Can be merged after core maintainer's review :)

**Review by ArthurZucker — APPROVED** [review]
Cool! Thanks for fixing!

**Pull request merged**
